test/lib/topology_builder.hh

```diff
@@ -0,0 +1,201 @@
+    }
+
+    void set_rack_id(int rack_id) {
+        _rack_id = rack_id;
+        _rack = fmt::format("rack{}", _rack_id);
+    }
+
+    // Locks the topology to prevent concurrent topology operations and load balancing.
+    // Setting transition_state to "lock" blocks background load-balancing which could interfere with the test
+    // and prevents errors from load_topology_state() complaining about nodes in transition with no transition state.
+    void lock_topology() {
+        abort_source as;
+        auto& client = _env.get_raft_group0_client();
+        service::topology& topo = _env.get_topology_state_machine().local()._topology;
+
+        while (true) {
+            if (topo.tstate && *topo.tstate == service::topology::transition_state::lock) {
+                testlog.info("Topology is locked");
+                return;
+            }
+
+            auto guard = client.start_operation(as).get();
+
+            if (topo.tstate) {
+                testlog.info("Waiting for topology state machine to be idle");
+                release_guard(std::move(guard));
+                _env.get_topology_state_machine().local().await_not_busy().get();
+                testlog.info("Woken up");
+                continue;
+            }
+
+            service::topology_change change({service::topology_mutation_builder(guard.write_timestamp())
+                                                 .set_transition_state(service::topology::transition_state::lock)
+                                                 .build()});
+            service::group0_command g0_cmd = client.prepare_command(std::move(change), guard, "locking topology");
+            try {
+                client.add_entry(std::move(g0_cmd), std::move(guard), as).get();
+            } catch (service::group0_concurrent_modification&) {
+                testlog.info("Concurrent modification detected while locking topology, retrying");
+            }
+        }
+    }
+public:
+    topology_builder(cql_test_env& e)
+        : _env(e)
+    {
+        start_new_dc();
+        lock_topology();
+    }
+
+    // Returns a new token from some sequence of unique tokens.
+    // Uniqueness is in the scope of the process, not just this object.
+    dht::token new_token() {
+        static int64_t next_token = 1;
+        return dht::token(next_token++);
+    }
+
+    // Returns the name of the currently built DC.
+    const sstring& dc() const {
+        return _dc;
+    }
+
+    // Returns location of the currently built rack.
+    endpoint_dc_rack rack() const {
+        return {_dc, _rack};
+    }
+
+    // Starts building a new rack in the current DC.
+    // Returns location of the new rack.
+    endpoint_dc_rack start_new_rack() {
+        set_rack_id(_rack_id + 1);
+        return rack();
+    }
+
+    // Starts building a new DC.
+    // DC is named uniquely in the scope of the process, not just this object.
+    void start_new_dc() {
+        static int next_id = 1;
```

Comment:
why not keep it as a class member?
We should prefer those over static variables

Reply:
So that each instance of `topology_builder` builds fresh DCs which don't alias to DCs created earlier in the same cql_test_env.

Reply:
Why is it better than a static counter? The name will be shorter than including a uuid in the name.

Reply:
If several test cases run in parallel, it would beneficial to tell which datacenters belong to the same cluster, but it is not crucial.